metaserver/src/main/scala/scala/meta/languageserver/search/SymbolIndex.scala

```diff
@@ -220,47 +267,81 @@ class SymbolIndex(
     }
   }
 
-  /** Returns a list of fallback symbols that can act instead of given symbol. */
-  private def alternatives(symbol: Symbol): List[Symbol] =
-    symbol match {
+  // TODO(alexey) this is not index specific and could be moved somewhere else (like Symbol-ops)
+  case object alternatives {
+
+    /** Returns a list of fallback symbols that can act instead of given symbol. */
+    // TODO(alexey) review this list
+    def forReferences(symbol: Symbol): List[Symbol] = {
+      List(
+        caseClassCompanionToType,
+        caseClassApplyOrCopyParams
+      ).flatMap { _.lift(symbol) }
+    }
+
+    /** Returns a list of fallback symbols that can act instead of given symbol. */
+    def forDefinition(symbol: Symbol): List[Symbol] = {
+      List(
+        caseClassCompanionToType,
+        caseClassApplyOrCopy,
+        caseClassApplyOrCopyParams,
+        methodToTerm
+      ).flatMap { _.lift(symbol) }
+    }
+
+    /** If `case class A(a: Int)` and there is no companion object, resolve
+     * `A` in `A(1)` to the class definition.
+     */
+    val caseClassCompanionToType: PartialFunction[Symbol, Symbol] = {
       case Symbol.Global(owner, Signature.Term(name)) =>
-        // If `case class A(a: Int)` and there is no companion object, resolve
-        // `A` in `A(1)` to the class definition.
-        Symbol.Global(owner, Signature.Type(name)) :: Nil
-      case Symbol.Multi(ss) =>
-        // If `import a.B` where `case class B()`, then
-        // resolve to either symbol, whichever has a definition.
-        ss
-      case Symbol.Global(
-          companion @ Symbol.Global(owner, signature),
-          Signature.Method("apply" | "copy", _)
-          ) =>
-        // If `case class Foo(a: Int)`, then resolve
-        // `apply` in `Foo.apply(1)`, and
-        // `copy` in `Foo(1).copy(a = 2)`
-        // to the `Foo` class definition.
-        companion :: Symbol.Global(owner, Signature.Type(signature.name)) :: Nil
```

Comment:
@olafurpg note that I removed this `companion`. Was it there for a good reason? If I'm not mistaken, this would redirect `Foo.apply` just to the companion object (if it doesn't have this `apply` defined explicitly).